### PR TITLE
Add initial support for tcp/ip sockets.

### PR DESCRIPTION
**Pull request by grant-arqit**
Initial draft to provide support for tcp/ip sockets.

### CODE REVIEW

File: tests/tcpsock.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+int iocb(struct davici_conn *c, int fd, int ops, void *user)
+{
+	assert(0);
+}
+
+int main(int argc, char *argv[])
+{
+	struct davici_conn *c;
```
Comment (mikemari82) [inline tests/tcpsock.c]: indentation

File: tests/tcpsock.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    pthread_t thread_id;
+    void *port = (void*) 55555;
+    pthread_create(&thread_id, NULL, mock_server, port);
+
+	int ret = davici_connect_tcpip("127.0.0.1:55555", iocb, NULL, &c);
```
Comment (mikemari82) [inline tests/tcpsock.c]: indentation

File: davici.c [hunk 3]
```diff
@@ -156,7 +127,6 @@ static int connect_in_and_fcntl(int fd, const char *path)
 
 	if(connect(fd, (struct sockaddr*)&addr, sizeof(addr)) < 0)
 	{
-		fprintf(stderr, "AF_NET connection failed.\n");
```
Comment (mikemari82) [inline davici.c]: Any reason to remove the fprintfs with error information?

Likewise for all fprintf's that have been removed

File: davici.c [hunk 4]
```diff
@@ -131,23 +104,21 @@ static void ip_and_port(char *ip_address, char *ip, char *port)
...
+static int connect_in_and_fcntl(int fd, const char *address)
 {
 	struct sockaddr_in addr;
 	int flags;
 
 	memset(&addr, 0, sizeof(addr));
 	addr.sin_family = AF_INET;
-	char *ip_address = malloc(strlen(path) * sizeof(char));
-	char *ip = malloc(strlen(path) * sizeof(char));
+	char *ip = malloc(strlen(address) * sizeof(char));
```
Comment (mikemari82) [inline davici.c]: this should be, likewise for others:

char *ip = malloc(strlen(address) + 1); 

To include the null terminator as strlen always returns the number of characters not including the null terminator, you technically don't need the sizeof(char) as it's always going to be 1 but can leave it for readability. 

I see the original code doesn't add the +1 which I think is a bug in their code

File: davici.c [hunk 5]
```diff
@@ -216,31 +186,50 @@ int davici_connect_unix(const char *path, davici_fdcb fdcb, void *user,
...
 	}
+	c->fdcb = fdcb;
+	c->user = user;
 
+	c->s = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (c->s < 0)
+	{
+		err = -errno;
+		free(c);
+		return err;
```
Comment (mikemari82) [inline davici.c]: can return "return -errno" to keep consistent with above, or vice versa

File: davici.h [hunk 6]
```diff
@@ -154,6 +154,28 @@ typedef int (*davici_recursecb)(struct davici_response *res, void *user);
...
+ * in theory is a blocking call.
+ *
+ * Using TCP could pose a security issue. Messages sent to it are not encrypted
+ * and so this should only be used in a controlled environment.
+ *
+ * @param address	address (ip:port) to TCP/IP socket
+ * @param fdcb		callback to register for file descriptor watching
+ * @param user		user context to pass to fdcb
+ * @param connp		pointer receiving connection context on success
+ * @return			0 on success, or a negative errno
```
Comment (mikemari82) [inline davici.h]: indentation

File: tests/badtcpsock.c [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	assert(0);
+}
+
+int main(int argc, char *argv[])
+{
+	struct davici_conn *c;
+	int ret;
+
+	ret = davici_connect_tcpip("127.0.0.1:66600", iocb, NULL, &c);
+    assert(ret != 0);
```
Comment (mikemari82) [inline tests/badtcpsock.c]: indentation

**Review by mikemari82 — COMMENTED** [review]
I did the commits in the wrong order, so excuse a few comments around fprintfs as I didn't realise you added them in

File: davici.c (hunk 3 above)
Comment (grant-arqit) [reply davici.c]: I had thought they shouldn't be in the production code, we should just return the errno. There are no other printf type statements in the existing code.

Comment (mikemari82) [reply davici.c]: I did the commits in the wrong order and thought the fprintfs were original code (i mentioned in an overall comment), so yes remove them

File: davici.c (hunk 4 above)
Comment (grant-arqit) [reply davici.c]: Have gone with your example.

**Comment by mikemari82** [comment]
Going to be a PITA but that formatting has put opening brackets on newlines, which I personally don't like. 

I think rather than:

void function
{
  // code
}

It should be:

void function {
  // code
}

Same goes for any opening bracket... like for if statements too. I think you would edit this in a VS code json file that defines the formatting rules

File: davici.c [hunk 8]
```diff
@@ -131,23 +104,21 @@ static void ip_and_port(char *ip_address, char *ip, char *port)
...
 {
 	struct sockaddr_in addr;
 	int flags;
 
 	memset(&addr, 0, sizeof(addr));
 	addr.sin_family = AF_INET;
-	char *ip_address = malloc(strlen(path) * sizeof(char));
-	char *ip = malloc(strlen(path) * sizeof(char));
+	char *ip = malloc(strlen(address) * sizeof(char));
+	char *ip_address = malloc(strlen(address) * sizeof(char));
```
Comment (mikemari82) [inline davici.c]: just thinking do you need this? Can you not pass address directly to `ip_and_port`?

**Comment by grant-arqit** [comment]
> Going to be a PITA but that formatting has put opening brackets on newlines, which I personally don't like.
> 
> I think rather than:
> 
> void function { // code }
> 
> It should be:
> 
> void function { // code }
> 
> Same goes for any opening bracket... like for if statements too. I think you would edit this in a VS code json file that defines the formatting rules

I do agree with you, however this is not the convention that the original author has gone with, so I'm reluctant to change that.

File: davici.c (hunk 8 above)
Comment (grant-arqit) [reply davici.c]: I need to separate out the `ip_address` and `port` from the overall `address` as the `ip_and_port` function uses the `:` as a token separator to set the `ip_address` and `port`

Comment (mikemari82) [reply davici.c]: but you create `ip` and `ip_address` and then pass `ip_address` to `ip_and_port`, just wondered if you could pass `address` directly to `ip_and_port` but just seen it takes a `char*` not a `const char*`

**Review by mikemari82 — APPROVED** [review]

**Pull request merged**